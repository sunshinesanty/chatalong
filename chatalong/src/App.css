.App {
  text-align: center;
}
.Chat {
  text-align: left;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  color: white;
}

.App-intro {
  font-size: large;
}
textarea {
  max-height: 75px;
  overflow: auto;
}
.chatRow .col-xs-10 {
    margin:5px 0 10px;
}
.uname {
  white-space: nowrap;
  height: 25px;
  padding-top: 6px;
}
.delContainer{
  text-align: right;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
