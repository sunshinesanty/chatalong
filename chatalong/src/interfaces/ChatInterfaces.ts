// tslint:disable:interface-name

import { ChatStoreObject } from '../Store/chatStore';
export interface IChatFormState {
    username: string;
    content: string;
}
export interface IStoreProp {
    chatStore?: ChatStoreObject;
}
export interface IPerChatProps extends IStoreProp {
    chatElement: IConversation;
}

export interface Iprops {
    chatStore?: ChatStoreObject;
}

export interface IChatFormProps extends Iprops {
    username: string;
    content: string;
    isDisabled: boolean;
    onSave: (ChatData: IChatFormState) => void;
}

export interface ILikeRendererProps extends Iprops {
    postID: number;
    username: string;
    likes: ILike[];
}
export interface ILikeRendererState {
    isLiked: boolean;
    likeCount: number;
}

export interface IConversation {
    id: number;
    text: string;
    username: string;
    created: string;
    modifed: string;
    likes: ILike[];
}
export interface ILike {
    id: number;
    concersationID: number;
    username: string;
    created: string;
    modified: string;
}

class ChatChaceKeys {
    get convs() { return 'convs'; }
}
export const ChaceKeys = new ChatChaceKeys();