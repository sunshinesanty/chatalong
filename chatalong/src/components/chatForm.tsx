import * as React from 'react';
import { IChatFormProps, IChatFormState } from '../interfaces/ChatInterfaces';
import { observer, inject } from 'mobx-react';

@inject('chatStore')
@observer
export class ChatForm extends React.Component<IChatFormProps, IChatFormState> {
    contentControl: HTMLTextAreaElement | null;
    // usernameControl: HTMLInputElement;
    constructor(props: IChatFormProps) {
        super(props);
        const { content, username } = this.props;
        this.state = {
            username,
            content
        };
    }
    onChange = (e: any) => {
        if (e.target.id.toLowerCase() === 'chatusername') {
            this.setState({ username: e.target.value });
            const { chatStore } = this.props;
            if (chatStore) {
                chatStore.activeUserName = e.target.value;
            } else {
                throw new ReferenceError('Chat Store not being passed as Property');
            }
        } else {
            this.setState({ content: e.target.value });
        }
    }
    onClick = (e: any) => {
        e.preventDefault();
        this.props.onSave(this.state);
        this.setState({ content: '' });
        if (this.contentControl) {
            this.contentControl.value = '';
        }
    }
    render() {
        return (
            <div className="container">
                <div className="form-group row">
                    <div className="col-12">                        
                        <input
                            type="text"
                            id="chatUserName"
                            value={this.state.username || ''}
                            onChange={this.onChange}
                            className="form-control"
                        />
                    </div>
                </div>
                <div className="form-group row">
                    <div className="col-12">
                        <textarea
                            id="chatContent"
                            cols={30}
                            rows={10}
                            onChange={this.onChange}
                            className="form-control"
                            defaultValue={this.state.content || ''}
                            ref={txtControl => this.contentControl = txtControl}
                        />
                    </div>
                </div>
                <div className="form-group row">
                    <div className="col-12">
                        <button
                            id="postBtn"
                            disabled={this.props.isDisabled}
                            style={(!this.props.isDisabled) ? { backgroundColor: 'grey' } : { backgroundColor: 'auto' }}
                            onClick={this.onClick}
                            title="add a comment then click to post"
                            className="btn btn-primary"
                        >
                            POST
                        </button>
                    </div>
                </div>
            </div>
        );
    }
}
