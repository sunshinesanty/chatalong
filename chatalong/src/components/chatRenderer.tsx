import * as React from 'react';

import { observer, inject } from 'mobx-react';
import PerChat from './perChat';
import { IConversation, Iprops } from '../interfaces/ChatInterfaces';
import LikeRenderer from './like';

@inject('chatStore')
@observer
export class ChatRenderer extends React.Component<Iprops, {}> {
    constructor(props: Iprops) {
        super(props);
        this.loadData();
    }

    loadData = async () => {
        if (this.props.chatStore) {
            await this.props.chatStore.chat.getAllConversations();
        } else {
            throw new ReferenceError('Chat Store not being passed as Property');
        }
    }

    render() {
        const { chatStore } = this.props;       
        let content;
        if (chatStore) {
            content = chatStore.chat.conversations.map(function (element: IConversation, idx: number) {
                return (
                    <PerChat key={element.id} chatElement={element} />
                );
            });
        }
        return (
            <div className="container Chat" >
                <ul className="list-group">
                    {content}
                </ul>
            </div>
        );
    }
}
