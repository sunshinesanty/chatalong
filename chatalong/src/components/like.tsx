import * as React from 'react';
import { ILikeRendererProps, ILikeRendererState } from '../interfaces/ChatInterfaces';
import { observer, inject } from 'mobx-react';

@inject('chatStore') @observer
class LikeRenderer extends React.Component<ILikeRendererProps, ILikeRendererState> {
    constructor(props: ILikeRendererProps) {
        super(props);
        const { username } = this.props;
        this.state = {
            isLiked: this.props.likes.findIndex(l =>
                l.username.toLowerCase() === username.toLowerCase()) > -1,
            likeCount: this.props.likes.length
        };
    }
    componentWillReceiveProps(nextProps: ILikeRendererProps) {
        const { chatStore } = this.props;
        if (chatStore) {
            chatStore.likes.isLikedByUser(nextProps.postID, nextProps.username).then((isLiked: boolean) => {
                this.setState({ isLiked });
            });
        } else {
            throw new ReferenceError('Chat Store not being passed as Property');
        }
    }
    onClicked = (e: any) => {
        e.preventDefault();
        let { isLiked, likeCount } = this.state;
        isLiked = !isLiked;
        likeCount = isLiked ? ++likeCount : --likeCount;
        this.setState({ isLiked, likeCount });
        const { chatStore } = this.props;
        if (chatStore) {
            chatStore.likes.toggleLike(this.props.postID, this.props.username);
        } else {
            throw new ReferenceError('Chat Store not being passed as Property');
        }
    }
    render() {
        const buttonStyle = `btn ${this.state.isLiked ? 'btn-info' : 'btn-default'} btn-sm`;
        const likeCountText = this.state.likeCount > 0 ? `(${this.state.likeCount})` : '';
        return (
            <div>
                <button type="button" className={buttonStyle} onClick={this.onClicked}>
                    <span className="glyphicon glyphicon-thumbs-up" /> {likeCountText}
                </button>
            </div>
        );
    }
}

export default LikeRenderer;
