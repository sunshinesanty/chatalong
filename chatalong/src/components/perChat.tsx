import * as React from 'react';
import { IPerChatProps } from '../interfaces/ChatInterfaces';
import { observer, inject } from 'mobx-react';
import LikeRenderer from './like';

@inject('chatStore') @observer
class PerChat extends React.Component<IPerChatProps, any> {
    render() {
        const { chatElement, chatStore } = this.props;
        const activeUserName = chatStore ? chatStore.activeUserName : '';
        return (
            <div>
                <li className="list-group-item">
                    <div className="row chatRow">
                        <div className="col-md-1 col-sm-2 col-xs-2 uname">{chatElement.username} : </div>
                        <div className="col-md-10 col-sm-8 col-xs-8">{chatElement.text}</div>
                        <div className="col-md-1 col-sm-2 col-xs-2 delContainer" >
                            <div>
                                <LikeRenderer
                                    postID={chatElement.id}
                                    likes={chatElement.likes}
                                    username={activeUserName}
                                />
                            </div>
                            <button type="button" className="btn btn-danger">Delete</button>
                        </div>
                    </div>
                </li>
            </div>
        );
    }
}

export default PerChat;
