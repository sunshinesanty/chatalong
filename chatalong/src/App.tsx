import * as React from 'react';
import './App.css';
import { ChatRenderer } from './components/chatRenderer';
import { ChatForm } from './components/chatForm';
import ChatStore from './Store/chatStore';
import DevTools from 'mobx-react-devtools';
import { Provider } from 'mobx-react';
import { IChatFormState, IConversation } from './interfaces/ChatInterfaces';
import { observer, inject } from 'mobx-react';
const logo = require('./logo.svg');

@inject('chatStore')
@observer
class App extends React.Component<any, any> {
  onSave = ({ username, content }: IChatFormState): void => {
    const conv: IConversation = {
      username,
      text: content,
      id: 0,
      likes: [],
      created: new Date().toISOString(),
      modifed: new Date().toISOString()
    };
    if (this.props.chatStore) {
      this.props.chatStore.chat.saveConversation(conv);
    }
  }
  render() {
    return (
      <Provider chatStore={ChatStore}>
        <div className="App container-fluid">
          <div className="App-header jumbotron">
            <img src={logo} className="App-logo" alt="logo" />
            <div className="page-header">
              <h2 className="">Welcome to React- Sample chat</h2>
            </div>
          </div>
          <DevTools />
          <ChatForm username="" content="" onSave={this.onSave} isDisabled={false} />
          <ChatRenderer />
        </div>
      </Provider>
    );
  }
}

export default App;
