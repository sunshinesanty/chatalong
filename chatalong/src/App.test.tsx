import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './App';
import { Provider } from 'mobx-react';
import ChatStore from './Store/chatStore';
import LikeRenderer from './components/like';
import { ChatRenderer } from './components/chatRenderer';
import PerChat from './components/perChat';
import * as renderer from 'react-test-renderer';
import commonUtils from './common/utils';
import { ChaceKeys, IConversation } from './interfaces/ChatInterfaces';
import DataCache from './api/storageHandler';

describe('general app test', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Provider chatStore={ChatStore}>
        <App />
      </Provider>,
      div);
  });
});

const initializeTestData = (generatedID: number, currDate: string) => {
  if (ChatStore.chat.conversations.length === 0) {
    ChatStore.activeUserName = 'Santy';
    ChatStore.chat.conversations.push({
      id: generatedID,
      created: currDate,
      modifed: currDate,
      likes: [{
        concersationID: generatedID,
        created: currDate,
        modified: currDate,
        id: commonUtils.generateID(),
        username: ChatStore.activeUserName
      }],
      text: 'Sample Test',
      username: ChatStore.activeUserName
    });
  }
  DataCache.setItem(ChaceKeys.convs, JSON.stringify(ChatStore.chat.conversations));
};

describe('Components rendering', () => {
  const generatedID = commonUtils.generateID();
  const currDate = new Date().toISOString();
  beforeEach(() => {
    initializeTestData(generatedID, currDate);
  });
  afterEach(() => {
    ChatStore.chat.conversations = [];
  });
  it('like should be rendered withut crashing', () => {
    const conv: IConversation | undefined = ChatStore.chat.conversations.find(c => c.id === generatedID); 
    const component = renderer.create(
      <Provider chatStore={ChatStore}>
        <LikeRenderer
          postID={generatedID}
          likes={conv ? conv.likes : []}
          username={ChatStore.activeUserName}
        />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('individual Chat should be rendered without crashing', () => {    
    const conv: any = ChatStore.chat.conversations.find(c => c.id === generatedID);
    const component = renderer.create(
      <Provider chatStore={ChatStore}>
        <PerChat chatElement={conv} />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('chat window to be rendered should be toggled', () => {
    const component = renderer.create(
      <Provider chatStore={ChatStore}>
        <ChatRenderer />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
