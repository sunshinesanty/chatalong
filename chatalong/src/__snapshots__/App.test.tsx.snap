// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components rendering chat window to be rendered should be toggled 1`] = `
<div
  className="container Chat"
>
  <ul
    className="list-group"
  />
</div>
`;

exports[`Components rendering individual Chat should be rendered without crashing 1`] = `
<div>
  <button
    className="btn btn-default btn-sm"
    onClick={[Function]}
    type="button"
  >
    <span
      className="glyphicon glyphicon-thumbs-up"
    />
     
    
  </button>
</div>
`;

exports[`Components rendering like should be rendered withut crashing 1`] = `
<div>
  <button
    className="btn btn-default btn-sm"
    onClick={[Function]}
    type="button"
  >
    <span
      className="glyphicon glyphicon-thumbs-up"
    />
     
    
  </button>
</div>
`;
